src/fsharp/FSharp.Core.Unittests/NameOfTests.fs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace FSharp.Core.Unittests
+open System
+open NUnit.Framework
+
+[<TestFixture>]
+type BasicNameOfTests() =    
+    let localConstant = 23
+    member this.MemberMethod() = 0 
+    member this.MemberProperty = this.MemberMethod()
+    static member StaticMethod() = 0 
+    static member StaticProperty = BasicNameOfTests.StaticMethod()
+        
+    [<Test>]
+    member this.``local variable name lookup`` () =
+        let a = 0
+        let result = nameof a
+        Assert.AreEqual("a",result)
+        Assert.AreEqual("result",nameof result)
+
+    [<Test>]
+    member this.``local int function name`` () =
+        let myFunction x = 0 * x
+        let b = nameof myFunction
+        Assert.AreEqual("myFunction",b)
+
+    [<Test>]
+    member this.``local curried function name`` () =
+        let curriedFunction x y = x * y
+        let b = nameof curriedFunction
+        Assert.AreEqual("curriedFunction",b)
+
+    [<Test>]
+    member this.``local tupled function name`` () =
+        let tupledFunction(x,y) = x * y
+        let b = nameof tupledFunction
+        Assert.AreEqual("tupledFunction",b)
+
+    [<Test>]
+    member this.``local unit function name`` () =
+        let myFunction() = 1
+        let b = nameof(myFunction)
+        Assert.AreEqual("myFunction",b)
+
+    [<Test>]
+    member this.``local function parameter name`` () =
+        let myFunction parameter1 = nameof parameter1
+
+        Assert.AreEqual("parameter1",myFunction "x")
+        Assert.AreEqual("parameter1",myFunction "x")
```

Comment:
We want to be really really sure here! 

Guess it's merge /rebase issue from long time ago